core/store/src/trie/split_state.rs

```diff
@@ -35,21 +40,110 @@ impl Trie {
+                TrieKey::DelayedReceipt { index } => match value {
+                    Some(value) => {
+                        let receipt = Receipt::try_from_slice(&value).map_err(|err| {
+                            StorageError::StorageInconsistentState(format!(
+                                "invalid delayed receipt {:?}, err: {}",
+                                value,
+                                err.to_string(),
+                            ))
+                        })?;
+                        insert_receipts.push((*index, receipt));
+                    }
+                    None => {}
+                },
+                TrieKey::Account { account_id }
+                | TrieKey::ContractCode { account_id }
+                | TrieKey::AccessKey { account_id, .. }
+                | TrieKey::ReceivedData { receiver_id: account_id, .. }
+                | TrieKey::PostponedReceiptId { receiver_id: account_id, .. }
+                | TrieKey::PendingDataCount { receiver_id: account_id, .. }
+                | TrieKey::PostponedReceipt { receiver_id: account_id, .. }
+                | TrieKey::ContractData { account_id, .. } => {
+                    let new_shard_uid = account_id_to_shard_id(account_id);
+                    // we can safely unwrap here because the caller of this function guarantees trie_updates contains all shard_uids for the new shards
+                    let trie_update = trie_updates.get_mut(&new_shard_uid).unwrap();
+                    match value {
+                        Some(value) => trie_update.set(trie_key, value),
+                        None => trie_update.remove(trie_key),
+                    }
+                }
+            }
+        }
+        for (_, update) in trie_updates.iter_mut() {
+            update.commit(StateChangeCause::Resharding);
+        }
+
+        insert_receipts.sort_by_key(|it| it.0);
+
+        let insert_receipts: Vec<_> =
+            insert_receipts.into_iter().map(|(_, receipt)| receipt).collect();
+
+        apply_delayed_receipts_to_split_states_impl(
+            &mut trie_updates,
+            &insert_receipts,
+            &changes.processed_delayed_receipts,
+            account_id_to_shard_id,
+        )?;
+
+        let mut trie_changes_map = HashMap::new();
+        for (shard_uid, update) in trie_updates {
+            let (trie_changes, _) = update.finalize()?;
+            trie_changes_map.insert(shard_uid, trie_changes);
+        }
+        Ok(trie_changes_map)
+    }
+
     /// Apply `changes` to build states for new shards
     /// `state_roots` contains state roots for the new shards
     /// The caller must guarantee that `state_roots` contains all shard_ids
     /// that `key_to_shard_id` that may return
     /// Ignore changes on DelayedReceipts or DelayedReceiptsIndices
     /// Update `store_update` and return new state_roots
     /// used for building states for new shards in resharding
```

Comment:
This comment is helpful, but I feel that it is not up to date:

```
/// Apply `changes` to build states for new shards
```
Does it refer to `apply_state_changes_to_split_states` above?

```
/// Update `store_update` and return new state_roots
/// used for building states for new shards in resharding
```
Should it be "Return `store_update` and new state_roots"?

Reply:
Good catch, I have updated the comments.

Thanks for pointing out! I think it is confusing that `apply_state_changes_to_split_states` and `add_values_to_split_states` are kind of similar. I have added comments for `apply_state_changes_to_split_states` as well. Hopefully that makes it clearer